nslsii/md_dict.py

```diff
@@ -14,15 +14,39 @@
 
 
 class RunEngineRedisDict(UserDict):
+    """
+    A class for storing RunEngine metadata added to RE.md on a Redis server.
+
+    This class has two strong ideas about the metadata it manages:
+      1. Some key-values are considered "global" or "facility-wide". These
+         are in use at all NSLS-II bluesky beamlines and include
+         proposal_id, data_session, cycle, SAF, and scan_id. The "global"
+         key-values are stored as Redis key-values. Redis does not support
+         numeric types, so the RunEngineRedisDict also keeps track of the
+         types of the "global" key-values. The intention is that this
+         metadata is accessible by any Redis client.
+      2. Non-global, or beamline-specific, metadata is stored in Redis as
+         a msgpack-ed blob. This means data type conversion between Redis
+         and Python is handled by msgpack, including numpy arrays.
+         The drawback is that this "local" metadata key-values are not
+         directly readable or writeable by Redis clients.
+    """
+
     PACKED_RUNENGINE_METADATA_KEY = "runengine-metadata-blob"
 
+    # these are the values used for global metadata
+    # if it does not exist when the RunEngineRedisDict
+    # is created
+    _missing_value = {str: "", int: int(-1), float: -1.0}
```

Comment:
Would it be better to allow the expected types to be `... or None` so that a missing value is an unambiguous out-of-band sentinel, rather than an in-band magic value? The branching logic that this leads to below looks potentially confusing. And it could it result in a field like `scan_id` being assigned `-1`?

Reply:
There were some complications here, and I landed on what seemed like an ok solution. Using msgpack with the "local" metadata makes this easy, but the "global" metadata is more difficult because those are pure Redis key-values. I think revisiting it now is a good idea.

Maybe a better solution than "default" global metadata values is just no default for those keys. Then RE.md raises KeyError and the user is made aware of the problem, which should happen very rarely.